### PR TITLE
New roles for smartcard server and client setup

### PR DESCRIPTION
**Pull request by t-woerner**
There are new smartcard roles in the roles folder:

    roles/ipasmartcard_server
    roles/ipasmartcard_client

This roles allows to setup smartcard for servers and clients.

Here is the documentation for the roles:

    roles/ipasmartcard_server/README.md
    roles/ipasmartcard_client/README.md

New example playbooks have been added:

    playbooks/install-smartcard-server.yml
    playbooks/install-smartcard-replicas.yml
    playbooks/install-smartcard-servers.yml
    playbooks/install-smartcard-clients.yml

### CODE REVIEW

File: roles/ipasmartcard_server/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    # If neither distro nor family is supported, try a default configuration.
+    - "vars/default.yml"
+
+- block:
+
+  # CA CERTS
+
+  # Fail on empty "ipasmartcard_server_ca_certs"
+  - name: Fail on empty "ipasmartcard_server_ca_certs"
+    fail: msg="No CA certs given in 'ipasmartcard_server_ca_certs'"
```
Comment (rjeffman) [inline roles/ipasmartcard_server/tasks/main.yml:29]: Some modules are using FQCN, some are not. I suggest using FQCN for all built-in modules.

The ones I found not using FQCN are `fail`, `package`, `command` and `script`.

File: roles/ipasmartcard_client/README.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
+ipasmartcard_client role
+========================
+
+Description
+-----------
+
+This role allows to configure IPA clients for Smart Card authentication.
+
+**Note**: The ansible playbook and role requires deployed IPA clients.
```
Comment (rjeffman) [inline roles/ipasmartcard_client/README.md:9]: Do the host have to be enrolled for this to work?

File: roles/ipasmartcard_client/README.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+FreeIPA versions 4.5 and up are supported by this role.
+
+
+Supported Distributions
+-----------------------
+
+* RHEL/CentOS 7.6+
+* Fedora 26+
+* Ubuntu
```
Comment (rjeffman) [inline roles/ipasmartcard_client/README.md:28]: I suggest removing "Ubuntu", as we'd need, at least, a vars file for it. For example, I haven't found the equivalent for `krb5-pkinit-openssl` package.

File: roles/ipasmartcard_client/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,173 @@
+---
+# tasks file for ipasmartcard_client role
+
+- name: Uninstall smartcard client
+  fail: msg="Uninstalling smartcard for IPA is not supported"
```
Comment (rjeffman) [inline roles/ipasmartcard_client/tasks/main.yml:5]: Some modules are using FQCN, some are not. I suggest using FQCN for all built-in modules.

The ones I found not using FQCN are `fail`, `package`, `command`, `set_fact` and `script`.

File: roles/ipasmartcard_server/README.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
+ipasmartcard_server role
+========================
+
+Description
+-----------
+
+This role allows to configure an IPA master or replica for Smart Card authentication.
+
+**Note**: The ansible playbooks and role require a deployed IPA server with ipa-ca.DOMAIN resolvable by the DNS server.
```
[the remark this answers is not in the document]
Comment (t-woerner) [reply roles/ipasmartcard_server/README.md:9]: Yes, ack.

File: roles/ipasmartcard_server/README.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+FreeIPA versions 4.5 and up are supported by this role.
+
+
+Supported Distributions
+-----------------------
+
+* RHEL/CentOS 7.6+
+* Fedora 26+
+* Ubuntu
```
[the remark this answers is not in the document]
Comment (t-woerner) [reply roles/ipasmartcard_server/README.md:29]: Yes, ack.

File: roles/ipasmartcard_server/README.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+Variables
+=========
+
+Variable | Description | Required
+-------- | ----------- | --------
+`ipaadmin_principal` | The kerberos principal used for admin. Will be set to `admin` if not set. (string) | no
+`ipaadmin_password` | The password for the IPA admin user. As an alternative an admin user keytab can be used instead with `ipaadmin_keytab`. (string) | yes
+`ipaadmin_keytab` | The admin keytab as an alternative to `ipaadmin_password`. (string) | no
+`ipaserver_hostname` | Fully qualified name of the server. By default ansible_facts['fqdn'] will be used. (string) | no
```
[the remark this answers is not in the document]
Comment (t-woerner) [reply roles/ipasmartcard_server/README.md:161]: Yes, ack.

File: roles/ipasmartcard_server/files/ipasmartcard_server_add_ca_to_sssd_auth_ca_db.sh [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash -eu
```
[the remark this answers is not in the document]
Comment (t-woerner) [reply roles/ipasmartcard_server/files/ipasmartcard_server_add_ca_to_sssd_auth_ca_db.sh:1]: O.k., will add them.

File: roles/ipasmartcard_server/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+    when: ipaadmin_password is defined
+
+  - name: kinit using "{{ ipaadmin_principal }}" keytab
+    command: kinit -kt "{{ ipaadmin_keytab }}" "{{ ipaadmin_principal }}"
+    when: ipaadmin_keytab is defined
+
+  # IS MASTER
+
+  - name: Check that this is an IPA master
+    command: ipa server-find --raw "{{ ipaserver_hostname | default(ansible_facts['fqdn']) }}"
```
[the remark this answers is not in the document]
Comment (t-woerner) [reply roles/ipasmartcard_server/tasks/main.yml:68]: Yes, ack.

File: roles/ipasmartcard_server/tasks/main.yml (hunk 1 above)
Comment (t-woerner) [reply roles/ipasmartcard_server/tasks/main.yml:29]: Yes, ack.

File: roles/ipasmartcard_client/README.md (hunk 2 above)
Comment (t-woerner) [reply roles/ipasmartcard_client/README.md:9]: Yes, the smartcard setup of the client requires a enrolled client as it is using kerberos and configured services on the client.

File: roles/ipasmartcard_client/README.md (hunk 3 above)
Comment (t-woerner) [reply roles/ipasmartcard_client/README.md:28]: Yes, ack.

**Review by rjeffman — APPROVED** [review]
LGTM

**Comment by varunmylaraiah** [comment]
LGTM, I hope that you will be adding upstream tests soon.

**Pull request merged**